1.3.0Alpha1 cannot initially find JDBC Driver on Java8, but works on Java11

**Describe the bug**
When using 1.3.0Alpha1 with Java8, during startup I see the following error:

```2020-02-07 08:14:19,847 INFO  [org.fly.cor.int.lic.VersionPrinter] (main) Flyway Community Edition 6.2.1 by Redgate
2020-02-07 08:14:19,867 WARN  [io.agr.pool] (Agroal_8663792271) Datasource '<default>': Unable to find a driver that accepts url: jdbc:postgresql://localhost:5432/__dbSchemaName__
2020-02-07 08:14:19,868 ERROR [io.qua.application] (main) Failed to start application: org.flywaydb.core.internal.exception.FlywaySqlException: 
Unable to obtain connection from database: Unable to find a driver that accepts url: jdbc:postgresql://localhost:5432/__dbSchemaName__
--------------------------------------------------------------------------------------------------------------------------------------
SQL State  : null
Error Code : 0
Message    : Unable to find a driver that accepts url: jdbc:postgresql://localhost:5432/__dbSchemaName__

	at org.flywaydb.core.internal.jdbc.JdbcUtils.openConnection(JdbcUtils.java:60)
	at org.flywaydb.core.internal.jdbc.JdbcConnectionFactory.<init>(JdbcConnectionFactory.java:80)
	at org.flywaydb.core.Flyway.execute(Flyway.java:447)
	at org.flywaydb.core.Flyway.migrate(Flyway.java:153)
	at io.quarkus.flyway.runtime.FlywayRecorder.migrate(FlywayRecorder.java:52)
	at io.quarkus.flyway.runtime.FlywayRecorder.doStartActions(FlywayRecorder.java:35)
	at io.quarkus.deployment.steps.FlywayProcessor$configureRuntimeProperties43.deploy_0(FlywayProcessor$configureRuntimeProperties43.zig:79)
	at io.quarkus.deployment.steps.FlywayProcessor$configureRuntimeProperties43.deploy(FlywayProcessor$configureRuntimeProperties43.zig:98)
	at io.quarkus.runner.ApplicationImpl.doStart(ApplicationImpl.zig:190)
	at io.quarkus.runtime.Application.start(Application.java:89)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at io.quarkus.runner.bootstrap.StartupActionImpl.run(StartupActionImpl.java:95)
	at io.quarkus.dev.IsolatedDevModeMain.firstStart(IsolatedDevModeMain.java:60)
	at io.quarkus.dev.IsolatedDevModeMain.accept(IsolatedDevModeMain.java:236)
	at io.quarkus.dev.IsolatedDevModeMain.accept(IsolatedDevModeMain.java:39)
	at io.quarkus.bootstrap.app.CuratedApplication.runInCl(CuratedApplication.java:132)
	at io.quarkus.bootstrap.app.CuratedApplication.runInAugmentClassLoader(CuratedApplication.java:86)
	at io.quarkus.dev.DevModeMain.start(DevModeMain.java:104)
	at io.quarkus.dev.DevModeMain.main(DevModeMain.java:46)
Caused by: java.sql.SQLException: Unable to find a driver that accepts url: jdbc:postgresql://localhost:5432/__dbSchemaName__
	at io.opentracing.contrib.jdbc.TracingDriver.findDriver(TracingDriver.java:132)
	at io.opentracing.contrib.jdbc.TracingDriver.connect(TracingDriver.java:68)
	at io.agroal.pool.ConnectionFactory.createConnection(ConnectionFactory.java:200)
	at io.agroal.pool.ConnectionPool$CreateConnectionTask.call(ConnectionPool.java:390)
	at io.agroal.pool.ConnectionPool$CreateConnectionTask.call(ConnectionPool.java:372)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at io.agroal.pool.util.PriorityScheduledExecutor.beforeExecute(PriorityScheduledExecutor.java:65)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1146)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

However, shortly after this error Quarkus attempts to load again and succeeds.

```
2020-02-07 08:14:19,909 INFO  [io.qua.dev.DevModeMain] (main) Attempting to start hot replacement endpoint to recover from previous Quarkus startup failure
2020-02-07 08:14:21,243 INFO  [io.qua.fly.FlywayProcessor] (build-22) Adding application migrations in path '/Users/sean.scott/dev/boreas/templates/quarkus-svc/src/main/g8/target/classes/db/migration/' using protocol 'file'
2020-02-07 08:14:21,244 INFO  [io.qua.fly.FlywayProcessor] (build-22) Adding application migrations in path '/Users/sean.scott/dev/boreas/templates/quarkus-svc/src/main/g8/target/classes/db/migration/' using protocol 'file'
2020-02-07 08:14:22,041 INFO  [io.qua.grpc] (vert.x-worker-thread-1) Registered gRPC service com.logrhythm.svc.__servicePackage__.v1.grpc.__serviceClass__GrpcService
2020-02-07 08:14:22,043 INFO  [io.qua.grpc] (vert.x-worker-thread-1) Registered gRPC interceptor com.logrhythm.core.quarkus.grpc.LRServerTracingInterceptor
2020-02-07 08:14:22,043 INFO  [io.qua.grpc] (vert.x-worker-thread-1) Registered gRPC interceptor com.logrhythm.core.quarkus.grpc.LRMethodInterceptor
2020-02-07 08:14:22,059 INFO  [io.qua.grpc] (vert.x-worker-thread-1) gRPC server listening on port 5050
2020-02-07 08:14:22,231 INFO  [org.fly.cor.int.dat.DatabaseFactory] (vert.x-worker-thread-1) Database: jdbc:postgresql://localhost:5432/__dbSchemaName__ (PostgreSQL 12.1)
2020-02-07 08:14:22,271 INFO  [org.fly.cor.int.com.DbMigrate] (vert.x-worker-thread-1) Current version of schema "public": 1
2020-02-07 08:14:22,272 INFO  [org.fly.cor.int.com.DbMigrate] (vert.x-worker-thread-1) Schema "public" is up to date. No migration necessary.
```

This does NOT happen in Java 11

**Expected behavior**
Error does not occur.

**Actual behavior**
Succeeds in Java 11, Fails in Java 8


**To Reproduce**
Steps to reproduce the behavior:
1. 
2. 
3. 

**Configuration**
```properties
# Add your application.properties here, if applicable.

```

**Screenshots**
(If applicable, add screenshots to help explain your problem.)

**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`: Darwin USMACEA090SSCOT 19.3.0 Darwin Kernel Version 19.3.0: Thu Jan  9 20:58:23 PST 2020; root:xnu-6153.81.5~1/RELEASE_X86_64 x86_64
 - Output of `java -version`: openjdk version "1.8.0_242"
OpenJDK Runtime Environment (AdoptOpenJDK)(build 1.8.0_242-b08)
OpenJDK 64-Bit Server VM (AdoptOpenJDK)(build 25.242-b08, mixed mode)
 - GraalVM version (if different from Java): 
 - Quarkus version or git rev: 1.3.0Alpha1

**Additional context**
(Add any other context about the problem here.)

This looks odd:

```
Caused by: java.sql.SQLException: Unable to find a driver that accepts url: jdbc:postgresql://localhost:5432/__dbSchemaName__
	at io.opentracing.contrib.jdbc.TracingDriver.findDriver(TracingDriver.java:132)
	at io.opentracing.contrib.jdbc.TracingDriver.connect(TracingDriver.java:68)
	at io.agroal.pool.ConnectionFactory.createConnection(ConnectionFactory.java:200)
```
Not sure why it's using `io.opentracing.contrib.jdbc.TracingDriver` to connect. Can you share an application which reproduces this?

--- a/extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/AbstractDataSourceProducer.java
+++ b/extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/AbstractDataSourceProducer.java
@@ -4,8 +4,10 @@ import java.sql.Connection;
 import java.sql.Driver;
 import java.sql.Statement;
 import java.util.ArrayList;
+import java.util.Iterator;
 import java.util.List;
 import java.util.Optional;
+import java.util.ServiceLoader;
 
 import javax.annotation.PreDestroy;
 import javax.inject.Inject;
@@ -73,6 +75,8 @@ public abstract class AbstractDataSourceProducer {
             log.warn("Datasource " + dataSourceName + " not started: driver and/or url are not defined.");
             return null;
         }
+        // we first make sure that all available JDBC drivers are loaded in the current TCCL
+        loadDriversInTCCL();
 
         DataSourceRuntimeConfig dataSourceRuntimeConfig = dataSourceRuntimeConfigOptional.get();
 
@@ -241,6 +245,24 @@ public abstract class AbstractDataSourceProducer {
         if (runtimeConfig == null) {
             throw new IllegalStateException(
                     "The datasources are not ready to be consumed: the runtime configuration has not been injected yet");
+        }
+    }
+
+    /**
+     * Uses the {@link ServiceLoader#load(Class) ServiceLoader to load the JDBC drivers} in context
+     * of the current {@link Thread#getContextClassLoader() TCCL}
+     */
+    private static void loadDriversInTCCL() {
+        // load JDBC drivers in the current TCCL
+        final ServiceLoader<Driver> drivers = ServiceLoader.load(Driver.class);
+        final Iterator<Driver> iterator = drivers.iterator();
+        while (iterator.hasNext()) {
+            try {
+                // load the driver
+                iterator.next();
+            } catch (Throwable t) {
+                // ignore
+            }
         }
     }
 

--- a/extensions/smallrye-opentracing/deployment/pom.xml
+++ b/extensions/smallrye-opentracing/deployment/pom.xml
@@ -75,6 +75,26 @@
             <artifactId>awaitility</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-hibernate-orm</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>io.opentracing.contrib</groupId>
+            <artifactId>opentracing-jdbc</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-agroal</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-jdbc-h2</artifactId>
+            <scope>test</scope>
+        </dependency>
     </dependencies>
 
     <build>

--- a/extensions/smallrye-opentracing/deployment/src/test/java/io/quarkus/smallrye/opentracing/deployment/Fruit.java
+++ b/extensions/smallrye-opentracing/deployment/src/test/java/io/quarkus/smallrye/opentracing/deployment/Fruit.java
@@ -1,0 +1,51 @@
+package io.quarkus.smallrye.opentracing.deployment;
+
+import javax.persistence.Cacheable;
+import javax.persistence.Column;
+import javax.persistence.Entity;
+import javax.persistence.GeneratedValue;
+import javax.persistence.GenerationType;
+import javax.persistence.Id;
+import javax.persistence.NamedQuery;
+import javax.persistence.QueryHint;
+import javax.persistence.SequenceGenerator;
+import javax.persistence.Table;
+
+@Entity
+@Table(name = "known_fruits")
+@NamedQuery(name = "Fruits.findAll", query = "SELECT f FROM Fruit f ORDER BY f.name", hints = @QueryHint(name = "org.hibernate.cacheable", value = "true"))
+@Cacheable
+public class Fruit {
+
+    @Id
+    @SequenceGenerator(name = "fruitsSequence", sequenceName = "known_fruits_id_seq", allocationSize = 1, initialValue = 10)
+    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "fruitsSequence")
+    private Integer id;
+
+    @Column(length = 40, unique = true)
+    private String name;
+
+    public Fruit() {
+    }
+
+    public Fruit(String name) {
+        this.name = name;
+    }
+
+    public Integer getId() {
+        return id;
+    }
+
+    public void setId(Integer id) {
+        this.id = id;
+    }
+
+    public String getName() {
+        return name;
+    }
+
+    public void setName(String name) {
+        this.name = name;
+    }
+
+}

--- a/extensions/smallrye-opentracing/deployment/src/test/java/io/quarkus/smallrye/opentracing/deployment/TracingTest.java
+++ b/extensions/smallrye-opentracing/deployment/src/test/java/io/quarkus/smallrye/opentracing/deployment/TracingTest.java
@@ -8,8 +8,8 @@ import org.jboss.shrinkwrap.api.ShrinkWrap;
 import org.jboss.shrinkwrap.api.asset.EmptyAsset;
 import org.jboss.shrinkwrap.api.spec.JavaArchive;
 import org.junit.jupiter.api.AfterAll;
-import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.api.extension.RegisterExtension;
 
@@ -30,6 +30,9 @@ public class TracingTest {
                     .addClass(TestResource.class)
                     .addClass(Service.class)
                     .addClass(RestService.class)
+                    .addClass(Fruit.class)
+                    .addAsResource("application.properties")
+                    .addAsResource("import.sql")
                     .addAsManifestResource(EmptyAsset.INSTANCE, "beans.xml"));
 
     static MockTracer mockTracer = new MockTracer();
@@ -37,7 +40,7 @@ public class TracingTest {
         GlobalTracer.register(mockTracer);
     }
 
-    @AfterEach
+    @BeforeEach
     public void after() {
         mockTracer.reset();
     }

--- a/extensions/smallrye-opentracing/deployment/src/test/resources/application.properties
+++ b/extensions/smallrye-opentracing/deployment/src/test/resources/application.properties
@@ -1,0 +1,9 @@
+quarkus.datasource.driver=io.opentracing.contrib.jdbc.TracingDriver
+quarkus.datasource.url=jdbc:tracing:h2:mem:test
+quarkus.datasource.max-size=8
+quarkus.datasource.min-size=2
+quarkus.hibernate-orm.database.generation=drop-and-create
+quarkus.hibernate-orm.log.sql=true
+quarkus.hibernate-orm.sql-load-script=import.sql
+quarkus.hibernate-orm.dialect=org.hibernate.dialect.H2Dialect
+

--- a/extensions/smallrye-opentracing/deployment/src/test/resources/import.sql
+++ b/extensions/smallrye-opentracing/deployment/src/test/resources/import.sql
@@ -1,0 +1,3 @@
+INSERT INTO known_fruits(id, name) VALUES (1, 'Cherry');
+INSERT INTO known_fruits(id, name) VALUES (2, 'Apple');
+INSERT INTO known_fruits(id, name) VALUES (3, 'Banana');

